### ISSUE TITLE
If the 'scalar' IO layer cant be loaded by the IO system Perl will create files on disk named after a reference.

### ISSUE THREAD

**Issue by demerphq** [Perl/perl5#21275]
Module:
PerlIO-scalar 

**Description**
If for some reason the PerlIO logic can't load the 'scalar' IO layer, then Perl will happily create files named after a reference on disk.

While working on https://github.com/Perl/perl5/pull/21274 and https://github.com/Perl/perl5/issues/21272 I discovered this.

**Steps to Reproduce**
This is a bit tricky. Configure a perl with -Dmksymlinks, do not install it, then write a script that does something like this:

```
chdir $out_of_the_symlink_tree;
open my $fh, ">", \$scalar or die "Failed to write to scalar buffer:$!";
print $fh "this is a test";
close $fh
```
where $out_of_the_symlink_tree is a directory not in the symlink tree created by -Dmksymlinks, for instance tmp. Run it from the symlink directory. It should produce an error like this:

```
Unknown PerlIO layer "scalar" at ...
```
and then it will leave a file in the directory like this: `SCALAR(0x559c23463250)`

**Expected behavior**
It should die during the open if the layer cant be found I think, or this Just Shouldn't Happen. Why do we load the scalar layer late?

**Perl configuration**
```
./perl -Ilib -V
Summary of my perl5 (revision 5 version 39 subversion 0) configuration:
  Local Commit: 286a358bf243d37f5e09f9ae3eba6c2e0239c65c
  Ancestor: ccfa6b533228ad41e4dbc5576cd43081b8fd2a13
  Platform:
    osname=linux
    osvers=5.15.0-76-generic
    archname=x86_64-linux-thread-multi
    uname='linux oncidium 5.15.0-76-generic #83~20.04.1-ubuntu smp wed jun 21 20:23:31 utc 2023 x86_64 x86_64 x86_64 gnulinux '
    config_args='-Dmksymlinks -des -Dusethreads -Doptimize=-g3 -ggdb3 -d -Dusedevel -Dcc=ccache gcc -Dld=gcc -DDEBUGGING'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=define
    usemultiplicity=define
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='gcc'
    ccflags ='-D_REENTRANT -D_GNU_SOURCE -fwrapv -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    optimize='-g3 -ggdb3'
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fwrapv -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion=''
    gccversion='9.4.0'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='gcc'
    ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/x86_64-linux-gnu /usr/lib /usr/lib64
    libs=-lpthread -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.31.so
    so=so
    useshrplib=false
    libperl=libperl.a
    gnulibc_version='2.31'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-Wl,-E'
    cccdlflags='-fPIC'
    lddlflags='-shared -g3 -ggdb3 -L/usr/local/lib -fstack-protector-strong'


Characteristics of this binary (from libperl): 
  Compile-time options:
    DEBUGGING
    HAS_LONG_DOUBLE
    HAS_STRTOLD
    HAS_TIMES
    MULTIPLICITY
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_HASH_FUNC_SIPHASH13
    PERL_HASH_USE_SBOX32
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    PERL_TRACK_MEMPOOL
    PERL_USE_DEVEL
    PERL_USE_SAFE_PUTENV
    USE_64_BIT_ALL
    USE_64_BIT_INT
    USE_ITHREADS
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_PERLIO
    USE_PERL_ATOF
    USE_REENTRANT_API
    USE_THREAD_SAFE_LOCALE
  Locally applied patches:
    286a358bf243d37f5e09f9ae3eba6c2e0239c65c
  Built under linux
  Compiled at Jul 20 2023 18:09:45
  %ENV:
    PERLBREW_CONFIGURE_FLAGS="-de -Dcc=ccache\ gcc -Dld=gcc"
    PERLBREW_HOME="/home/yorton/.perlbrew"
    PERLBREW_MANPATH="/home/yorton/perl5/perlbrew/perls/perl-5.34.1/man"
    PERLBREW_PATH="/home/yorton/perl5/perlbrew/bin:/home/yorton/perl5/perlbrew/perls/perl-5.34.1/bin"
    PERLBREW_PERL="perl-5.34.1"
    PERLBREW_ROOT="/home/yorton/perl5/perlbrew"
    PERLBREW_SHELLRC_VERSION="0.88"
    PERLBREW_VERSION="0.88"
  @INC:
    lib
    /usr/local/lib/perl5/site_perl/5.39.0/x86_64-linux-thread-multi
    /usr/local/lib/perl5/site_perl/5.39.0
    /usr/local/lib/perl5/5.39.0/x86_64-linux-thread-multi
    /usr/local/lib/perl5/5.39.0
```

**Comment by Leont** [comment]
We should probably just move `PerlIO::scalar` into core, I have no idea why it is a (core-only) extension

**Comment by demerphq** [comment]
> We should probably just move `PerlIO::scalar` into core

Good idea. I see there are four such libraries:

```
$ ls ext | grep PerlIO
PerlIO-encoding
PerlIO-mmap
PerlIO-scalar
PerlIO-via
```
I am not sure what is involved with moving these into core. Could we instead make it so that they get loaded automatically by universal.c? So long as they are loaded earlier these kinds of problem shouldn't happen I think. What say you @Leont ?

**Comment by Leont** [comment]
The other ones aren't loaded as magically as :scalar is, so the case for moving them is less strong. And TBH I'm not sure why :mmap is in core in the first place.

**Comment by tonycoz** [comment]
Fixed by #21282
